### PR TITLE
Code specialization

### PR DESCRIPTION
**Pull request by AndrasKovacs**
Decrease `-foptimise-heavily` build time by 15-17% by adding a bunch of `SPECIALIZE` pragmas. Also add a bunch of `INLINE` and `INLINABLE` pragmas. `-foptimise-heavily` runtime performance is unchanged within margin of error. 

Pragmas are placed based on the following information:
- Module build times displayed with https://github.com/codedownio/time-ghc-modules and `-ddump-timings -ddump-to-file` GHC options. 
- Core printed with `-ddump-simpl -dsuppress-all -dno-suppress-type-signatures -ddump-to-file`.
- Failed specializations printed with `-Wall-missed-specializations` when we build without `-foptimise-heavily`. 

Only the worst offending and most visible instances of code duplication are addressed here; there is definitely a large amount that's still there.

### CODE REVIEW

File: src/full/Agda/Syntax/Translation/InternalToAbstract.hs [hunk 1]
```diff
@@ -408,14 +417,15 @@ instance Reify Literal where
...
   reifyWhen = reifyWhenE
-  reify l = return $ A.Lit empty l
+  reify l = return $ A.Lit empty l; {-# INLINE reify #-}
 
 instance Reify Term where
   type ReifiesTo Term = Expr
 
   reifyWhen = reifyWhenE
-  reify v = reifyTerm True v
+  reify v = reifyTerm True v; {-# inline reify #-}
```
Comment (andreasabel) [inline src/full/Agda/Syntax/Translation/InternalToAbstract.hs]: INLINE

File: src/full/Agda/Interaction/BasicOps.hs [hunk 2]
```diff
@@ -426,80 +425,82 @@ reifyElimToExpr = \case
...
+    (a,(x,b)) <- reify (unDom a,b)
+    return $ PTSInstance a b
+  reify (CheckDataSort q s) = DataSort q <$> reify s
+  reify (CheckLockedVars t _ lk _) = CheckLock <$> reify t <*> reify (unArg lk)
+  reify (CheckMetaInst m) = do
+    t <- jMetaType . mvJudgement <$> lookupLocalMeta m
+    OfType <$> reify (MetaV m []) <*> reify t
+  reify (CheckType t) = JustType <$> reify t
+  reify (UsableAtModality _ _ mod t) = UsableAtMod mod <$> reify t
+  {-# SPECIALIZE reify :: Constraint -> TCM (ReifiesTo Constraint) #-}
```
Comment (andreasabel) [inline src/full/Agda/Interaction/BasicOps.hs:498]: There is a large diff just fixing some indentation.  Could whitespace changes be put into a separate commit please? (can then be ignored in a git blame).

File: src/full/Agda/TypeChecking/Pretty.hs [hunk 3]
```diff
@@ -161,36 +166,55 @@ prettyTCMCtx :: (PrettyTCM a, MonadPretty m) => Precedence -> a -> m Doc
...
+instance PrettyTCM Bool                       where prettyTCM = pretty
+instance PrettyTCM C.Name                     where prettyTCM = pretty
+instance PrettyTCM C.QName                    where prettyTCM = pretty
+instance PrettyTCM TopLevelModuleName         where prettyTCM = pretty
+instance PrettyTCM Comparison                 where prettyTCM = pretty
+instance PrettyTCM Literal                    where prettyTCM = pretty
+instance PrettyTCM Nat                        where prettyTCM = pretty
+instance PrettyTCM ProblemId                  where prettyTCM = pretty
+instance PrettyTCM Range                      where prettyTCM = pretty
+instance PrettyTCM CheckpointId               where prettyTCM = pretty
```
Comment (andreasabel) [inline src/full/Agda/TypeChecking/Pretty.hs:178]: Please put this whitespace change into a separate commit.

File: src/full/Agda/Syntax/Position.hs [hunk 4]
```diff
@@ -735,7 +737,7 @@ x `withRangeOf` y = setRange (getRange y) x
 --   ending position is placed first. If both tie, the element from the
 --   first list is placed first.
 interleaveRanges :: (HasRange a) => [a] -> [a] -> ([a], [(a,a)])
-interleaveRanges as bs = runWriter$ go as bs
+interleaveRanges as bs = runWriter $ go as bs
```
Comment (andreasabel) [inline src/full/Agda/Syntax/Position.hs:740]: whitespace

File: src/full/Agda/Syntax/Translation/AbstractToConcrete.hs [hunk 5]
```diff
@@ -1,4 +1,4 @@
-{-# LANGUAGE CPP          #-}
+{-# LANGUAGE CPP #-}
```
Comment (andreasabel) [inline src/full/Agda/Syntax/Translation/AbstractToConcrete.hs:1]: whitespace

File: src/full/Agda/Interaction/BasicOps.hs [hunk 6]
```diff
@@ -1,5 +1,4 @@
 {-# LANGUAGE NondecreasingIndentation #-}
-
```
Comment (andreasabel) [inline src/full/Agda/Interaction/BasicOps.hs:2]: whitespace

File: src/full/Agda/TypeChecking/CheckInternal.hs [hunk 7]
```diff
@@ -219,8 +226,8 @@ checkArgInfo action ai ai' = do
   mod <- checkModality action (getModality ai)  (getModality ai')
   return $ setModality mod ai
 
-checkHiding    :: (MonadCheckInternal m) => Hiding -> Hiding -> m ()
-checkHiding    h h' = unless (sameHiding h h') $ typeError $ HidingMismatch h h'
+checkHiding :: (MonadCheckInternal m) => Hiding -> Hiding -> m ()
+checkHiding h h' = unless (sameHiding h h') $ typeError $ HidingMismatch h h'
```
Comment (andreasabel) [inline src/full/Agda/TypeChecking/CheckInternal.hs:230]: whitespace

**Review by andreasabel — COMMENTED** [review]
Thanks, this is great!

One request for commit hygiene:  Could you please gather the whitespace changes into a whitespace-only commit?  This is good when coming back to these commits in the future, e.g. via blame or bisect.

**Comment by AndrasKovacs** [comment]
I moved the whitespace edits to a commit.

File: src/full/Agda/TypeChecking/Monad/MetaVars.hs [hunk 8]
```diff
@@ -774,15 +775,13 @@ solveAwakeConstraints' = solveSomeAwakeConstraints (const True)
...
+freezeMetas :: forall m. MonadTCState m => LocalMetaStore -> m (Set MetaId)
 freezeMetas ms =
   execWriterT $
   modifyTCLensM stOpenMetaStore $
   execStateT (mapM_ freeze $ MapS.keys ms)
   where
-  freeze ::
-    Monad m =>
-    MetaId -> StateT LocalMetaStore (WriterT (Set MetaId) m) ()
+  freeze :: MetaId -> StateT LocalMetaStore (WriterT (Set MetaId) m) ()
```
Comment (andreasabel) [inline src/full/Agda/TypeChecking/Monad/MetaVars.hs:784]: Borderline whitespace change :-D (not really just whitespace, but ok).

File: src/full/Agda/TypeChecking/Monad/Base.hs [hunk 9]
```diff
@@ -3820,9 +3828,7 @@ eQuantity f e =
     | otherwise      = __IMPOSSIBLE__
 
 eHardCompileTimeMode :: Lens' TCEnv Bool
-eHardCompileTimeMode f e =
-  f (envHardCompileTimeMode e) <&>
-  \x -> e { envHardCompileTimeMode = x }
+eHardCompileTimeMode f e = f (envHardCompileTimeMode e) <&> \x -> e { envHardCompileTimeMode = x }
```
Comment (andreasabel) [inline src/full/Agda/TypeChecking/Monad/Base.hs:3833]: Looks like a whitespace change.  Oh well.

**Review by andreasabel — APPROVED** [review]
Thanks for the update!

The sorting into ws/non-ws is now 99% correct, let's ignore the 2 outliers.

File: src/full/Agda/TypeChecking/Pretty.hs [hunk 10]
```diff
@@ -175,22 +180,37 @@ instance PrettyTCM CheckpointId where prettyTCM = pretty
...
+{-# SPECIALIZE prettyTCM :: Bool               -> TCM Doc #-}
+{-# SPECIALIZE prettyTCM :: C.Name             -> TCM Doc #-}
+{-# SPECIALIZE prettyTCM :: C.QName            -> TCM Doc #-}
+{-# SPECIALIZE prettyTCM :: TopLevelModuleName -> TCM Doc #-}
+{-# SPECIALIZE prettyTCM :: Comparison         -> TCM Doc #-}
+{-# SPECIALIZE prettyTCM :: Literal            -> TCM Doc #-}
+{-# SPECIALIZE prettyTCM :: Nat                -> TCM Doc #-}
+{-# SPECIALIZE prettyTCM :: ProblemId          -> TCM Doc #-}
+{-# SPECIALIZE prettyTCM :: Range              -> TCM Doc #-}
+{-# SPECIALIZE prettyTCM :: CheckpointId       -> TCM Doc #-}
```
Comment (andreasabel) [inline src/full/Agda/TypeChecking/Pretty.hs:192]: Why did these functions get SPECIALIZE rather then INLINE pragmas?
(Converse question below.)

File: src/full/Agda/TypeChecking/Pretty.hs [hunk 11]
```diff
@@ -274,48 +319,54 @@ instance PrettyTCM Elim where
...
   prettyTCM = prettyTCM . ignoreReduced
+{-# SPECIALIZE prettyTCM :: PrettyTCM a => MaybeReduced a -> TCM Doc #-}
 
 instance PrettyTCM EqualityView where
   prettyTCM v = prettyTCM $ equalityUnview v
+{-# SPECIALIZE prettyTCM :: EqualityView -> TCM Doc #-}
 
 instance PrettyTCM A.Expr where
-  prettyTCM = prettyA
+  prettyTCM = prettyA; {-# INLINE prettyTCM #-}
```
Comment (andreasabel) [inline src/full/Agda/TypeChecking/Pretty.hs:337]: Why did these functions get INLINE pragmas rather than SPECIALIZE ones?

Comment (andreasabel) [reply src/full/Agda/TypeChecking/Pretty.hs:337]: @AndrasKovacs answered that SPECIALIZE for all would be safer.
